### PR TITLE
Major changes

### PR DESCRIPTION
**Pull request by MorGuux**
Hi there,

I found your mod after an excellent demo from @Ultrawipf. It's cool to see how the free camera can be made into a drone with some simple physics applied, and the results are great. I can see some professional drone pilots creating some nice videos of virtual drifting with it.

I've had a go at re-writing the physics it uses, in order to allow users to input real data about the drone. This should allow the simulated physics to correlate with the real thing (hopefully). I also added some new input options (inverted axis, combined axis) to allow users (like myself) to use a gamepad, rather than a flight controller. I also fixed some issues regarding the flight dynamics being affected by the framerate. I've not written anything in Python before, so it was a personal challenge of learning a new language too. It also taught me about AC app development, which is nice.

Have a look and see what you think. I've tested it personally and had others too, and we think the gravity is still a bit weak, though I don't want to artificially increase it since the drag equations and gravity forces should all be true-to-life. Maybe I'm missing something obvious in the simulation that may solve it.

I've changed the default config (with the new format) to some tested values that feel decent, but I guess it depends on the type of drones you fly and what you're comfortable with. Feel free to experiment and report back on what you think of it 😄

Morgan Gardner

## Major Changes
* Physics re-write, now use mass, air density, wingspan (calculated surface area), and forces calculated for gravity, drag, and thrust.
* Input scaling altered, separating from deltaT (old system was affected by framerate).
* config.defaults.ini added. Used by Content Manager to provide descriptions, defaults, and new controls to the app settings

### CODE REVIEW

**Comment by dot1nt** [comment]
Thanks so much. There is still some weird stuff going on with the physics but i will look into it as soon as i have time.

File: apps/python/acfpv/drone.py [hunk 1]
```diff
@@ -68,32 +76,46 @@ def yaw(self, val):
...
-    def move(self):
+        new_accel = [0, 0, 0]
+        new_velocity = [0, 0, 0]
 
-        pos = list(self.position)
-        pos[2] += self.x
-        pos[0] += self.y
-        pos[1] += self.z
+        for a in range(len(self.position)):
+            new_position[a] = self.position[a] + self.velocity[a] * deltaT + self.gravity * (deltaT * deltaT * 0.5)
```
Comment (dot1nt) [inline apps/python/acfpv/drone.py]:
```suggestion
            new_position[a] = self.position[a] + self.velocity[a] * deltaT
```

File: apps/python/acfpv/drone.py [hunk 2]
```diff
@@ -68,32 +76,46 @@ def yaw(self, val):
...
+            force_drag[a] = drag((config.drag / 100), self.surface_area, self.air_density, self.velocity[a])
+            force_throttle[a] = self.throttle_mag[a] * (config.power_to_weight * -force_gravity)
+
+        force_total[0] = -force_drag[0] + force_throttle[0]
+        force_total[1] = -force_drag[1] + -force_throttle[1]
+        force_total[2] = force_gravity - force_drag[2] + force_throttle[2]
+
+        for a in range(len(self.position)):
+            new_accel[a] = force_total[a] / (config.mass / 1000)
+            new_velocity[a] = self.velocity[a] + (self.acceleration[a] + new_accel[a]) * (deltaT * 0.5)
```
Comment (dot1nt) [inline apps/python/acfpv/drone.py]:
```suggestion
            new_velocity[a] = self.velocity[a] + (self.acceleration[a] + new_accel[a]) * deltaT

```

**Review by dot1nt — COMMENTED** [review]
This seems to fix it. Why did you add gravity to the new position twice and multiplied the acceleration with half the deltaT?

**Comment by MorGuux** [comment]
Honestly, it was not my equation that I used, I found it somewhere, but I agree, looking at it now, it doesn't match the basic physics equations for acceleration, velocity, etc.

After testing your changes I do see that it feels much better (the drone drops faster with gravity).

A final thing to mention may be to add some rotational inertia, so the forces involved in rotating the drone (giving the wingspan and mass) are applied, rather than directly controlling the roll/pitch/yaw from the joystick inputs directly.

**Comment by dot1nt** [comment]
Yes, that would be an idea. I also thought about adding wind, different surface area depending on the rotation or a full pid controller. I will merge this now. Thanks

**Comment by MorGuux** [comment]
What about converting to quaternions instead of Euler rotations? I'm sure I noticed before that there's a slight drift off-axis (not noticeable while actually flying around). I was testing the gravity and drag by limiting the motion to just vertical Z, and I started getting ever so slight X and Y movement too. Maybe it's not related, but it's definitely not an issue in the real use-case.

**Comment by dot1nt** [comment]
I noticed that too but only on some maps (i think). I don't think quaternions would fix that and working with them is probably more difficult. I also didn't have a problem with gimbal lock while testing so i don't know if it's worth it.

**Pull request merged**